app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java

```diff
@@ -0,0 +1,81 @@
+package com.codesoom.assignment.controllers;
+
+import com.codesoom.assignment.application.UserService;
+import com.codesoom.assignment.domain.User;
+import com.codesoom.assignment.dto.UserCreateRequest;
+import com.codesoom.assignment.dto.UserUpdateRequest;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+import static org.mockito.BDDMockito.given;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@DisplayName("UserController 클래스")
+@WebMvcTest(UserControllerTest.class)
```

Comment:
```suggestion
@WebMvcTest(UserController.class)
```
이렇게 변경하시면 해결될거에요!

댓글에 써있듯이 이 코드에 어노테이션에 대해서 심도있게 찾아보시고 공유해주세요!

Reply:
넵 잘 알겠습니다. 도움주셔서 감사드립니다!

## `@WebMvcTest()`
 
- `@Controller`, `@ControllerAdvice` 등의 빈을 스캔한다. [자동 설정 목록](https://docs.spring.io/spring-boot/docs/current/reference/html/test-auto-configuration.html#appendix.test-auto-configuration)
- `MockMvc`로 컨트롤러를 자동 설정한다.
	- 하나의 컨트롤러로 제한하지 않으려면 `@MockBean`으로 등록한다.

```java
@Autowired  
private MockMvc mockMvc;
  
@MockBean  
private UserService userService;
```
 
### 장점
- 웹 애플리케이션 관련 빈만 등록하므로 빠르다.
-  통합 테스트를 진행하기 어려운 테스트를 진행 가능하다.
ex) 결제 모듈 API를 콜하며 안되는 상황에서 Mock을 통해 가짜 객체를 만들어 테스트 가능.

### 단점
 - 요청부터 응답까지 모든 테스트를 Mock 기반으로 테스트하기 때문에 실제 환경에서는 제대로 동작하지 않을 수 있다.

### 예제

```java
@Test  
@DisplayName("찾은 사용자와 상태코드 200을 응답한다")  
void it_responds_found_user_and_status_code_200() throws Exception {  
    //given  
    given(userService.getUser(EXISTING_ID)).willReturn(user);  
    //when  
    mockMvc.perform(get("/users/{id}", EXISTING_ID)  
                    .accept(MediaType.APPLICATION_JSON)  
                    .contentType(MediaType.APPLICATION_JSON))  
              
            // then  
            .andExpect(jsonPath("$.id").exists())  
            .andExpect(jsonPath("name").exists())  
            .andExpect(jsonPath("email").exists())  
            .andExpect(status().isOk());  
}
```

- given: Mock 객체 생성
- when: Mock 객체 반환
- then: 응답 검사

## 출처
- [sprint boot doc](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.testing.spring-boot-applications.spring-mvc-tests)
- [갓대희의 작은공간:티스토리](https://goddaehee.tistory.com/212)

Reply:
좋아요!